src/osd/OSD.cc

```diff
@@ -8725,6 +8751,34 @@ void OSD::enqueue_op(PG *pg, OpRequestRef& op)
 	   << " cost " << op->get_req()->get_cost()
 	   << " latency " << latency
 	   << " " << *(op->get_req()) << dendl;
+
+  // 15 --> op lock completion
+  if (op->get_req()->get_type() == MSG_OSD_REPOPREPLY) {
+    MOSDRepOpReply *r = static_cast<MOSDRepOpReply *>(op->get_req());
+    r->finish_decode();
+    spg_t pgid = pg->get_pgid();
+    const pg_pool_t* pp = osdmap->get_pg_pool(pgid.pool());
+    if (r->get_result() == MSG_OSD_OP_COMP_RESULT && 
+	pp->get_type() == pg_pool_t::TYPE_REPLICATED) {
+      
+
+      PGBackend * pg_backend = pg->get_pgbackend();
+      if (!pg_backend) {
+	dout(0) << __func__ << " PGBackend is null " << dendl;
+	pg->unlock();
+	assert(0);
+      }
+
+      ReplicatedBackend * rep_backend = dynamic_cast<ReplicatedBackend*>(pg_backend);
+      if (!rep_backend) {
+	dout(0) << __func__ << " ReplicatedBackend is null " << dendl;
+	pg->unlock();
+	assert(0);
+      }
+      rep_backend->sub_op_modify_reply_no_pg_lock(op);
```

Comment:
Right. I thought sending ack in an ms_fast_dispath cause no overhead to messenger code because it can be processed fast. otherwise sending sharded work queue cause consuming additional CPU.